### PR TITLE
manager: Enhance event filtering for performance

### PR DESCRIPTION
**Pull request by gtjoseph**
UserNote: You can now perform more granular filtering on events
in manager.conf using expressions like
`eventfilter(name(Newchannel),header(Channel),method(starts_with)) = PJSIP/`
This is much more efficient than
`eventfilter = Event: Newchannel.*Channel: PJSIP/`
Full syntax guide is in configs/samples/manager.conf.sample.

Also split the XML documentation to a separate file as it was almost 1500 lines.

### CODE REVIEW

**Comment by gtjoseph** [comment]
cherry-pick-to: 18
cherry-pick-to: 20
cherry-pick-to: 21
cherry-pick-to: 22

**Comment by gtjoseph** [comment]
If you can think of more test scenarios to run, let me know.

File: main/manager_doc.xml [hunk 1]
```diff
@@ -0,0 +1,1623 @@
...
+				<para>
+				Advanced match criteria.  If not specified, the <literal>Filter</literal>
+				parameter is assumed to be a regular expression and will be matched against
+				the entire event payload.
+				</para>
+				<para>
+				Syntax: [name(&lt;event_name&gt;)][,header(&lt;header_name&gt;)][,&lt;match_method&gt;]
+				</para>
+				<para>
+				One each of the following may be specified separated by commas.
```
Comment (bkford) [inline main/manager_doc.xml:1333]: One of each

File: main/manager_doc.xml [hunk 2]
```diff
@@ -0,0 +1,1623 @@
...
+				<para>This filter option is used to whitelist or blacklist events per user to be
+				reported with regular expressions and are allowed if both the regex matches
+				and the user has read access as defined in manager.conf. Filters are assumed to be for whitelisting
+				unless preceeded by an exclamation point, which marks it as being black.
+				Evaluation of the filters is as follows:</para>
+				<para>- If no filters are configured all events are reported as normal.</para>
+				<para>- If there are white filters only: implied black all filter processed first, then white filters.</para>
+				<para>- If there are black filters only: implied white all filter processed first, then black filters.</para>
+				<para>- If there are both white and black filters: implied black all filter processed first, then white
+				filters, and lastly black filters.</para>
```
Comment (bkford) [inline main/manager_doc.xml:1409]: include/exclude list

**Review by jcolp — CHANGES_REQUESTED** [review]
Marking request per Ben comments, nothing else noted so will approve once those are resolved

File: main/manager_doc.xml (hunk 2 above)
Comment (gtjoseph) [reply main/manager_doc.xml]: I removed this section entirely and replaced it with...

> See the manager.conf.sample file in the configs/samples directory of the Asterisk source tree for a full description and examples.

**Review by jcolp — APPROVED** [review]

File: configs/samples/manager.conf.sample [hunk 3]
```diff
@@ -104,22 +104,184 @@ bindaddr = 0.0.0.0
...
+; value of the Channel header this is still more efficient than using the
+; legacy method.
+
+; Only include Newchannel and Hangup events whose "Channel" header doesn't
+; start with Local/.
+;eventfilter(action(include),name(Newchannel)) =
+;eventfilter(action(include),name(Hangup)) =
+;eventfilter(header(Channel),action(exclude),method(starts_with)) = Local/
+; No regexes at all.  We do the hash match against the event names first and
+; only mathcing events are passed to the next filter.
```
Comment (henningw) [inline configs/samples/manager.conf.sample:266]: spelling error, mathcing -> matching

Comment (gtjoseph) [reply configs/samples/manager.conf.sample:266]: Thanks.  I'll fix it shortly.

**Pull request merged**